pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go

```diff
@@ -468,108 +470,180 @@ func panicIfColReferencedInPredicate(b BuildCtx, col *scpb.Column, tblElts Eleme
+	return haveSameIndexColsByKind(b, tableID, indexID1, indexID2, scpb.IndexColumn_KEY) &&
+		haveSameIndexColsByKind(b, tableID, indexID1, indexID2, scpb.IndexColumn_KEY_SUFFIX) &&
+		haveSameIndexColsByKind(b, tableID, indexID1, indexID2, scpb.IndexColumn_STORED)
+}
+
+// getAllPrimaryIndexesSortedBySourcing returns all "adding" primary indexes
+// in the table and ensure they are "sorted by sourcing".
+//   - "adding" means they are targeting public but currently not public;
+//   - "sorted by sourcing" means if primary_index_j's source index ID points
+//     to primary_index_i, then primary_index_i comes before primary_index_j.
+//
+// We conclude that the return has at least 1 and at most 4 primary indexes.
+// To facilitate conversation, let's call them (`old`, `inter1`, `inter2`, `final`):
+//   - `old` is the original, currently public primary index (it's always going
+//     to exist and hence "at least 1").
+//   - `inter1` is the newly added primary index that contains all the added/dropped
+//     columns in this statement.
+//   - `inter2` is same as `inter1` but with altered primary key.
+//   - `final` is same as `inter2` but without dropped column.
+//
+// The following comments explain in what cases would we have 2, 3, or 4 adding
+// primary indexes.
+//
+// Usually, if there is just one add column, or one drop column, or one
+// alter primary key in one alter table statement, we will only create one new
+// primary index with the correct columns. That would be `final` and we backfill
+// it from `old`, that is, (`old`, nil, nil, `final`).
+//
+// Occasionally, we might need one intermediate primary index. It happens in the
+// following two cases:
+// 1). `ALTER TABLE t ALTER PRIMARY KEY where old PK is on rowid;`
+// 2). `ALTER TALBE t ADD COLUMN, DROP COLUMN;`
+// For 1), the intermediate primary index will be one with the altered PK but
+// retaining rowid in its storing columns. The final primary index will then be
+// one that drops rowid from its storing columns, so, (`old`, nil, `inter2`, `final`).
+// For 2), the intermediate primary index will be one with all the added and
+// dropped columns in its storing columns. Its final primary index will then be
+// one that drops those to-be-dropped columns from its storing columns, so,
+// (`old`, `inter1`, nil, `final`).
+//
+// Rarely, we would encounter something like
+// `ALTER TABLE ADD COLUMN, DROP COLUMN, ALTER PRIMARY KEY;`
+// To correctly build this statement, we will need two intermediate primary
+// indexes where intermediate1 will be one that has all the added and dropped
+// columns in its storing columns, and intermediate2 will be one that drops all
+// to-be-dropped columns from its storing columns, and `final` will be one with
+// altered PK, so, (`old`, `inter1`, `inter2`, `final`).
+func getAllPrimaryIndexesSortedBySourcing(
+	b BuildCtx, tableID catid.DescID,
+) (old, inter1, inter2, final *scpb.PrimaryIndex) {
```

Comment:
My impression reading a lot of the code in this package is that you might want some kind of struct instead of this tuple, and define methods on it.